README.rst

.. image:: https://github.com/xlcnd/isbnlib/workflows/tests/badge.svg
    :target: https://github.com/xlcnd/isbnlib/actions
    :alt: Built Status

.. image:: https://img.shields.io/github/issues/xlcnd/isbnlib/bug.svg?label=bugs&style=flat
    :target: https://github.com/xlcnd/isbnlib/labels/bug
    :alt: Bugs

.. image:: https://img.shields.io/pypi/dm/isbnlib.svg?style=flat
    :target: https://pypi.org/project/isbnlib/
    :alt: PYPI Downloads



Info
====

``isbnlib`` is a (pure) python library that provides several
useful methods and functions to validate, clean, transform, hyphenate and
get metadata for ISBN strings.



Install
-------

From the command line, enter (in some cases you have to precede the
command with ``sudo``):


.. code-block:: bash

    $ pip install isbnlib


If you use linux systems, you can install using your distribution package
manager (all major distributions have packages ``python-isbnlib``
and ``python3-isbnlib``), however (usually) are **very old and don't work well anymore**!



ISBN
----

   The official form of an ISBN is something like ``ISBN 979-10-90636-07-1``. However for most
   applications only the numbers are important, you can always 'mask' them if you need (see below).
   This library works mainly with 'striped' ISBNs  (only digits and X) like '0826497527'. You can
   strip an ISBN-like string by using ``canonical(isbnlike)``. You can
   'mask' the ISBN by using ``mask(isbn)``. So in the examples below, when you see 'isbn'
   in the argument, it is a 'striped' ISBN, when the argument is an 'isbnlike' it is a string
   like ``ISBN 979-10-90636-07-1`` or even something dirty like ``asdf 979-10-90636-07-1 bla bla``.

   Two important concepts: **valid ISBN** should be an ISBN that was built according with the rules,
   this is distinct from **issued ISBN** that is an ISBN that was already issued to a publisher
   (this is the usage of the libraries and most of the web services).
   However *isbn.org*, probably by legal reasons, merges the two!
   So, according to *isbn-international.org*, '9786610326266' is not valid (because the block 978-66...
   has not been issued yet, however if you use ``is_isbn13('9786610326266')`` you will get ``True``
   (because '9786610326266' follows the rules of an ISBN). But the situation is even murkier,
   try ``meta('9786610326266')`` and you will see that this ISBN was already used!

   If possible, work with ISBNs in the isbn-13 format (since 2007, only are issued ISBNs
   in the isbn-13 format). You can always convert isbn-10 to isbn-13, but **not** the reverse (read this_).
   Read more about ISBN at isbn-international.org_ or wikipedia_.



Main Functions
--------------

``is_isbn10(isbn10like)``
    Validates as ISBN-10.

``is_isbn13(isbn13like)``
    Validates as ISBN-13.

``to_isbn10(isbn13)``
    Transforms isbn-13 to isbn-10.

``to_isbn13(isbn10)``
    Transforms isbn-10 to isbn-13.

``canonical(isbnlike)``
    Keeps only digits and X. You will get strings like `9780321534965` and `954430603X`.

``clean(isbnlike)``
    Cleans ISBN (only legal characters).

``notisbn(isbnlike, level='strict')``
    Check with the goal to invalidate isbn-like.

``get_isbnlike(text, level='normal')``
    Extracts all substrings that seem like ISBNs (very useful for scraping).

``get_canonical_isbn(isbnlike, output='bouth')``
    Extracts ISBNs and transform them to the canonical form.

``ean13(isbnlike)``
    Transforms an `isbnlike` string into an EAN13 number (validated canonical ISBN-13).

``doi(isbn)``
    Returns a DOI's ISBN-A from a ISBN-13.

``mask(isbn, separator='-')``
    `Mask` (hyphenate) a canonical ISBN.

``info(isbn)``
    Gets the language or country assigned to this ISBN.

``meta(isbn, service='default')``
    Gives you the main metadata associated with the ISBN. As `service` parameter you can use:
    ``'goob'`` uses the **Google Books service** (**no key is needed**)  and
    **is the default option**,
    ``'wiki'`` uses the **wikipedia.org** api (**no key is needed**),
    ``'openl'`` uses the **OpenLibrary.org** api (**no key is needed**).
    You can enter API keys
    with ``config.add_apikey(service, apikey)`` (see example below).
    The output can be formatted as ``bibtex``, ``csl`` (CSL-JSON), ``msword``, ``endnote``, ``refworks``,
    ``opf`` or ``json`` (BibJSON) bibliographic formats with ``registry.bibformatters``.
    Now, you can extend the functionality of this function by adding plugins, more metadata
    providers or new bibliographic formatters (check_ for available plugins).

``editions(isbn, service='merge')``
    Returns the list of ISBNs of editions related with this ISBN. By default
    uses 'merge' (merges 'openl', 'thingl' and 'wiki'), but other providers are available:
    'openl' (uses the search API from **Open Library**),
    'thingl' (uses the service ThingISBN from **LibraryThing**),
    'wiki' (uses the service Citation from **Wikipedia**)
    and 'any' (first tries 'wiki', if no data then 'openl').

``isbn_from_words(words)``
    Returns the most probable ISBN from a list of words (for your geographic area).

``goom(words)``
    Returns a list of references from **Google Books multiple references**.

``classify(isbn)``
    Returns a dictionary of **classifiers** for a canonical ISBN. For the meaning of these classifiers see OCLC_.
    Most of the data in the underlying service are for books in english.

``desc(isbn)``
    Returns a small description of the book.
    *Almost all data available are for US books!*

``cover(isbn)``
    Returns a dictionary with the url for cover.
    *Almost all data available are for US books!*

``doi2tex(DOI)``
    Returns metadata formatted as BibTeX for a given DOI.

``ren(filename)``
    Renames a file using metadata from an ISBN in his filename.


See files test_core_ and test_ext_ for **a lot of examples**.



Plugins
-------

You can extend the functionality of the library by adding plugins (for now, just
new metadata providers or new bibliographic formatters).

For available plugins check_ here.

After install, your plugin will blend transparently in ``isbnlib`` (you will have more options in ``meta`` and ``bibformatters``).




For Devs
========


API's Main Namespaces
---------------------

In the namespace ``isbnlib`` you have access to the **core functions**:
``is_isbn10``, ``is_isbn13``, ``to_isbn10``, ``to_isbn13``, ``canonical``,
``clean``, ``notisbn``, ``get_isbnlike``, ``get_canonical_isbn``, ``mask``,
``info``, ``check_digit10``, ``check_digit13``, ``doi`` and ``ean13``.

In addition, you have access to **metadata functions**, namely:
``meta``, ``editions``, ``ren``, ``desc``, ``cover``,
``goom``, ``classify``, ``doi2tex`` and ``isbn_from_words``.

The exceptions raised by these methods can all be caught using ``ISBNLibException``.


You can extend the lib by using the classes and functions exposed in
namespace ``isbnlib.dev``, namely:

* ``WEBService`` a class that handles the access to web
  services (just by passing an url) and supports ``gzip``.
  You can subclass it to extend the functionality... but
  probably you don't need to use it! It is used in the next class.

* ``WEBQuery`` a class that uses ``WEBService`` to retrieve and parse
  data from a web service. You can build a new provider of metadata
  by subclassing this class.
  His main methods allow passing custom
  functions (*handlers*) that specialize them to specific needs (``data_checker`` and
  ``parser``). It implements a **throttling mechanism** with a default rate of
  one call per second per service.

* ``Metadata`` a class that structures, cleans and 'validates' records of
  metadata. His method ``merge`` allows to implement a simple merging
  procedure for records from different sources. The main features of this class, can be
  implemented by a call to the ``stdmeta`` function instead!

* ``vias`` exposes several functions to put calls to services, just by passing the name and
  a pointer to the service's ``query`` function.
  ``vias.parallel`` allows to put threaded calls.
  You can use ``vias.serial`` to make serial calls and
  ``vias.multi`` to use several cores. The default is ``vias.serial``.

The exceptions raised by these methods can all be caught using ``ISBNLibDevException`` (or, more general, ``ISBNLibException``).
You **shouldn't raise** this exception in your code, only raise the specific exceptions
exposed in ``isbnlib.dev`` whose name ends in Error.


In ``isbnlib.dev.helpers`` you can find several methods, that we found very useful, some of then
are only used in ``isbntools`` (*an app and framework* that uses ``isbnlib``).


With ``isbnlib.config`` you can read and set configuration options:
change timeouts with ``seturlopentimeout`` and ``setthreadstimeout``,
access api keys with ``apikeys`` and add new one with ``add_apikey``,
access and set generic and user-defined options with ``options.get('OPTION1')`` and ``set_option``.


Finally, from ``isbnlib.registry`` you can change the metadata service to be used by default
(``setdefaultservice``),
add a new service (``add_service``), access bibliographic formatters for metadata (``bibformatters``),
set the default formatter (``setdefaultbibformatter``), add new formatters (``add_bibformatter``) and
set a new cache (``set_cache``) (e.g. to switch off the cache ``set_cache(None)``).
The cache only works for calls through metadata functions. These changes only work for the 'current session',
so should be done always before calling other methods.


Let us concretize these points with a small example.

Suppose you want a small script to get metadata using ``Open Library`` formatted in BibTeX.

A minimal script would be:


.. code-block:: python

    from isbnlib import meta
    from isbnlib.registry import bibformatters

    SERVICE = "openl"

    # now you can use the service
    isbn = "9780446310789"
    bibtex = bibformatters["bibtex"]
    print(bibtex(meta(isbn, SERVICE)))



Patterns of Usage
-----------------

The library implements a very simple API with sensible defaults, but there are cases
that need your attention (see case 3 below).



A. You only need **core functions**:


.. code-block:: python

    # import the core functions you need
    from isbnlib import canonical, is_isbn10, is_isbn13

    isbn = canonical("978-0446310789")
    if is_isbn13(isbn):
        ...
    ...


B. You need also **metadata functions**, with **default config**:


.. code-block:: python

    from isbnlib import canonical, meta, description

    isbn = canonical("978-0446310789")
    data = meta(isbn)
    ...

C. You need also **metadata functions**, with **special config**:

   *Lets suppose you need to add an api key for a metadata plugin
   and change the cache too*.


.. code-block:: python

    from myapp.utils import MyCache

    # import the functions you need, plus 'config' and 'registry'
    from isbnlib import canonical, config, meta, registry

    # you should use 'config' first
    config.add_apikey("isbndb", "kjshdfkjahsdflkjh")

    # then 'registry'
    registry.set_cache(MyCache())

    # Only now you should use metadata functions
    # (there are no adaptions for core functions,
    #  so they can be used at any moment)
    isbn = canonical("978-0446310789")
    data = meta(isbn, service="isbndb")
    ...


D. You want to build a **plugin** or use **isbnlib.dev** in your code:

   You should study very carefully the **public** methods in ``dir(isbnlib.dev)``, start with this template_
   and follow the instructions there. For inspiration take a look at goob_.

   Most of the public bibliographic catalog services return data in **SRU** or **Unimarc** format. It is very easy
   to write a customer **plugin** for these services, just use porbase_ (SRU) or sbn_ (Unimarc) as templates
   and consult this project_.



Caveats
-------


1. These classes are optimized for one-call to services and not for batch calls.

2. If you inspect the library, you will see that there are a lot of private modules
   (their name starts with '_'). These modules **should not** be accessed directly since,
   with high probability, your program will break with a further version of the library!



Projects using *isbnlib*
========================

**Open Library**   https://github.com/internetarchive/openlibrary

**NYPL Library Simplified**  https://github.com/NYPL-Simplified

**RERO ILS**  https://github.com/rero/rero-ils

**CERN CDS RDM** https://github.com/CERNDocumentServer/cds-rdm

**ResearchHub** https://github.com/ResearchHub/researchhub-backend

**Manubot**   https://github.com/manubot

**isbntools**      https://github.com/xlcnd/isbntools

**isbnsrv**        https://github.com/xlcnd/isbnsrv



See the full list here_.



Help
====


If you need help, please take a look at github_ or post a question on
stackoverflow_ .



.. _github: https://github.com/xlcnd/isbnlib/discussions

.. _range: https://www.isbn-international.org/range_file_generation

.. _isbntools: https://pypi.python.org/pypi/isbntools

.. _sourcegraph: http://bit.ly/ISBNLib_srcgraph

.. _readthedocs: http://bit.ly/ISBNLib_rtd

.. _stackoverflow: http://stackoverflow.com/search?tab=newest&q=isbnlib

.. _test_core: https://github.com/xlcnd/isbnlib/blob/master/isbnlib/test/test_core.py

.. _test_ext: https://github.com/xlcnd/isbnlib/blob/master/isbnlib/test/test_ext.py

.. _isbn-international.org: https://www.isbn-international.org/content/what-isbn

.. _wikipedia: http://en.wikipedia.org/wiki/International_Standard_Book_Number

.. _python-future.org: http://python-future.org/compatible_idioms.html

.. _issue: https://github.com/xlcnd/isbnlib/issues/28

.. _check: https://pypi.python.org/pypi?%3Aaction=search&term=isbnlib_&submit=search

.. _template: https://github.com/xlcnd/isbnlib/blob/dev/PLUGIN.zip

.. _goob: https://github.com/xlcnd/isbnlib/blob/dev/isbnlib/_goob.py

.. _search: https://pypi.python.org/pypi?%3Aaction=search&term=isbnlib&submit=search

.. _51: https://github.com/xlcnd/isbnlib/issues/51

.. _here: https://github.com/xlcnd/isbnlib/network/dependents?package_id=UGFja2FnZS01MjIyODAxMQ%3D%3D

.. _OCLC: http://classify.oclc.org/classify2/

.. _this: https://bisg.org/news/479346/New-979-ISBN-Prefixes-Expected-in-2020.htm

.. _sbn: https://github.com/arangb/isbnlib-sbn/blob/main/isbnlib_sbn/_sbn.py

.. _porbase: https://github.com/xlcnd/isbnlib-porbase/blob/dev/isbnlib_porbase/_porbase.py

.. _project: https://github.com/xlcnd/isbnlib/issues?q=is%3Aissue+is%3Aopen+label%3Aproject


CHANGES.txt
v3.3.3, 2014-06-16 -- Initial release (based in `isbntools 3.3.3)`.
v3.3.4, 2014-06-16 -- Fix small bug with wheel package (based in `isbntools 3.3.3)`.
v3.3.5, 2014-06-22 -- Fix bug in 'notisbn' (based in `isbntools 3.3.4)`.
v3.3.6, 2014-06-30 -- Small improvements and optional ShelveCache.
v3.3.7, 2014-07-07 -- Fix issue #2, better debug logging and expose top exceptions.
v3.3.8, 2014-07-07 -- Fix wrong logging identifier.
v3.3.9, 2014-07-13 -- Mark as not valid ISBN-13 in series not allocated yet.
v3.4.1, 2014-07-16 -- Fix bug #7 and small improvements on the API.
v3.4.2, 2014-08-13 -- Add `doi2tex` and updated data.
v3.4.3, 2015-01-13 -- Better LOOSE regex and allow 'x' in canonical.
v3.4.4, 2015-01-26 -- Helpers sprint and fake_isbn.
v3.4.5, 2015-02-05 -- Better yield for openl and better log model.
v3.4.6, 2015-02-05 -- Fix bug #13.
v3.4.7, 2015-02-11 -- Fix bug #14.
v3.4.8, 2015-02-25 -- Fix bug #16, add 'desc' and 'cover' features.
v3.4.9, 2015-02-27 -- Fix issue #17 and bug #18, data range 20150227.
v3.5.1, 2015-03-10 -- New CoversCache and better unicode printing in Windows.
v3.5.2, 2015-03-15 -- Fix bug #19, relative paths on coverscache and data range 20150310.
v3.5.3, 2015-03-26 -- Add throttling for web services, data range 20150325 and 'in memory' keys cache.
v3.5.4, 2015-04-10 -- Issue #20, fix bugs #21 and #22 and data range 20150401.
v3.5.5, 2015-04-22 -- Add logging to vias, fix bug #23, and data range 20150422.
v3.5.6, 2015-06-03 -- Deprecated 'bouth23' and data range 20150603.
v3.5.7, 2015-11-22 -- Support for py35 and pypy3 and data range 20151118.
v3.5.8, 2016-03-07 -- Issue #28 (closing down of xISBN service).
v3.5.9, 2016-06-17 -- Improved data quality (close #39 and #40) and new cover (close issue #42).
v3.6.1, 2016-06-21 -- Improved data quality (#39 and #40), new cover (#42) and streamlined.
v3.6.2, 2016-09-06 -- Updated data.
v3.6.3, 2016-11-09 -- Updated data.
v3.6.4, 2016-11-12 -- Fix bug #46.
v3.6.5, 2016-12-08 -- Updated data.
v3.6.6, 2017-01-05 -- Fix bug #47.
v3.6.7, 2017-03-11 -- Updated data.
v3.6.8, 2017-03-31 -- New functions: check_digit10 and check_digit13.
v3.7.1, 2017-05-24 -- Fix #43 (bump middle version).
v3.7.2, 2017-06-19 -- Restore xID (improved data quality close #28).
v3.7.3, 2018-01-08 -- Updated data.
v3.7.4, 2018-01-08 -- Add support for CSL-JSON format (close #48).
v3.8.1, 2018-01-24 -- BREAK: stop support for py26, py33 and isbndb.
v3.8.2, 2018-01-24 -- Fix bug 'in_virtual'.
v3.8.3, 2018-01-29 -- Add 'wcat' to editions.
v3.8.4, 2018-03-06 -- Solve issues with non-ASCII searches in 'from_words'.
v3.8.5, 2018-06-14 -- NOT RELEASED.
v3.9.1, 2018-07-02 -- Issue #51 (closing down of xISBN service).
v3.9.2, 2018-09-28 -- Updated data.
v3.9.3, 2018-10-05 -- Make it easier to override 'WEBService.data'.
v3.9.4, 2019-01-18 -- Fix bibjson and updated data.
v3.9.5, 2019-02-10 -- PR #53 and updated data.
v3.9.6, 2019-02-24 -- Performance improvements (close #55).
v3.9.7, 2019-05-06 -- Editions is now cached and consistent return types.
v3.9.8, 2019-05-06 -- Fix bug #58 on 'editions'.
v3.9.9, 2019-10-17 -- Better caching, improved security, fix 'catch all exception'.
v3.9.10, 2019-12-20 -- Fix issues: 61 (close #61) and 62 (close #62).
v3.9.11, 2019-12-20 -- NOT RELEASED.
v3.10.0, 2020-03-23 -- Add a 'classify' service (oclc.org).
v3.10.1, 2020-05-18 -- Add a 'wiki' (wikipedia) provider for metadata and editions.
v3.10.2, 2020-05-26 -- Fix 'goob' and 'wiki' quirks (issue #64 and #65).
v3.10.3, 2020-05-27 -- Delete a (quick debug) print statement.
v3.10.4, 2020-11-17 -- Add 'Isbn' and fix #67.
v3.10.5, 2020-12-31 -- Fix 'csv' and updated data.
v3.10.6, 2021-01-26 -- Fix 'wiki' (close #75).
v3.10.7, 2021-04-12 -- Common url for GB services.
v3.10.8, 2021-05-10 -- New 'Authors' parsing strategy for 'wiki'.
v3.10.9, 2021-11-01 -- Fix 'editions'.
v3.10.10, 2022-02-25 -- Fix issue #87.
v3.10.11, 2022-10-18 -- Fix issues #107 #114.
v3.10.12, 2022-10-18 -- Fix tests on 'master'.
v3.10.13, 2023-02-15 -- Fix issues #119, #121 and updated data.
v3.10.14, 2023-04-26 -- Fix issue #123.
v3.11.0, 2023-04-26 -- TENTATIVE


isbnlib/_core.py
# -*- coding: utf-8 -*-

# isbnlib - tools for extracting, cleaning and transforming ISBNs
# Copyright (C) 2014-2023 Alexandre Lima Conde
# SPDX-License-Identifier: LGPL-3.0-or-later

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""isbnlib main file for ISBN manipulation.

Tools for extracting, cleaning, transforming and validating ISBN ids.
"""

import logging
import re

LOGGER = logging.getLogger(__name__)

RE_ISBN10 = re.compile(r'ISBN\x20(?=.{13}$)\d{1,5}([- ])\d{1,7}'
                       r'\1\d{1,6}\1(\d|X)$|[- 0-9X]{10,16}')
RE_ISBN13 = re.compile(r'97[89]{1}(?:-?\d){10,16}|97[89]{1}[- 0-9]{10,16}')
RE_STRICT = re.compile(
    r'^(?:ISBN(?:-1[03])?:? )?(?=[0-9X]{10}$|'
    r'(?=(?:[0-9]+[- ]){3})'
    r'[- 0-9X]{13}$|97[89][0-9]{10}$|'
    r'(?=(?:[0-9]+[- ]){4})'
    r'[- 0-9]{17}$)(?:97[89][- ]?)?[0-9]{1,5}'
    r'[- ]?[0-9]+[- ]?[0-9]+[- ]?[0-9X]$',
    re.I | re.M | re.S,
)
RE_NORMAL = re.compile(
    r'97[89]{1}-?[0-9]{10}|'
    r'97[89]{1}-[-0-9]{13}|'
    r'\d{9}[0-9X]{1}|'
    r'[-0-9X]{10,16}',
    re.I | re.M | re.S,
)
RE_LOOSE = re.compile(r'[- 0-9X]{10,19}', re.I | re.M | re.S)
ISBN13_PREFIX = '978'  # just the prefix with isbn-10
LEGAL = '0123456789xXisbnISBN- '


def check_digit10(firstninedigits):
    """Check sum ISBN-10."""
    # minimum checks
    if len(firstninedigits) != 9:
        return ''
    try:
        int(firstninedigits)
    except ValueError:  # pragma: no cover
        return ''
    # checksum
    val = sum(
        (i + 2) * int(x) for i, x in enumerate(reversed(firstninedigits)))
    remainder = int(val % 11)
    if remainder == 0:
        tenthdigit = 0
    else:
        tenthdigit = 11 - remainder
    if tenthdigit == 10:
        tenthdigit = 'X'
    return str(tenthdigit)


def check_digit13(firsttwelvedigits):
    """Check sum ISBN-13."""
    # minimum checks
    if len(firsttwelvedigits) != 12:
        return ''
    try:
        int(firsttwelvedigits)
    except ValueError:  # pragma: no cover
        return ''
    # checksum
    val = sum(
        (i % 2 * 2 + 1) * int(x) for i, x in enumerate(firsttwelvedigits))
    thirteenthdigit = 10 - int(val % 10)
    if thirteenthdigit == 10:
        thirteenthdigit = '0'
    return str(thirteenthdigit)


def _check_structure10(isbn10like):
    """Check structure of an ISBN-10."""
    return bool(re.match(RE_ISBN10, isbn10like))


def _check_structure13(isbn13like):
    """Check structure of an ISBN-13."""
    return bool(re.match(RE_ISBN13, isbn13like))


def is_isbn10(isbn10):
    """Validate as ISBN-10."""
    isbn10 = canonical(isbn10)
    if len(isbn10) != 10:
        return False  # pragma: no cover
    return bool(not check_digit10(isbn10[:-1]) != isbn10[-1])


def is_isbn13(isbn13):
    """Validate as ISBN-13."""
    isbn13 = canonical(isbn13)
    if len(isbn13) != 13:
        return False  # pragma: no cover
    if isbn13[0:3] not in ('978', '979'):
        return False
    return bool(not check_digit13(isbn13[:-1]) != isbn13[-1])


def to_isbn10(isbn13):
    """Transform isbn-13 to isbn-10."""
    isbn13 = canonical(isbn13)
    # Check prefix
    if isbn13[:3] != ISBN13_PREFIX:
        return isbn13 if len(isbn13) == 10 and is_isbn10(isbn13) else ''
    if not is_isbn13(isbn13):
        return ''
    isbn10 = isbn13[3:]
    check = check_digit10(isbn10[:-1])
    # Change check digit
    return isbn10[:-1] + check if check else ''


def to_isbn13(isbn10):
    """Transform isbn-10 to isbn-13."""
    isbn10 = canonical(isbn10)
    if len(isbn10) == 13 and is_isbn13(isbn10):
        return isbn10
    if not is_isbn10(isbn10):
        return ''
    isbn13 = ISBN13_PREFIX + isbn10[:-1]
    check = check_digit13(isbn13)
    return isbn13 + check if check else ''


def canonical(isbnlike):
    """Keep only numbers and X."""
    numb = [c for c in isbnlike if c in '0123456789Xx']
    if numb and numb[-1] == 'x':
        numb[-1] = 'X'
    isbn = ''.join(numb)
    # Filter some special cases
    if (isbn and len(isbn) not in (10, 13)
            or isbn in ('0000000000', '0000000000000', '000000000X')
            or isbn.find('X') not in (9, -1) or isbn.find('x') != -1):
        return ''
    return isbn


def clean(isbnlike):
    """Clean ISBN (only legal characters)."""
    cisbn = [c for c in isbnlike if c in LEGAL]
    buf = re.sub(r'\s*-\s*', '-', ''.join(cisbn))
    return re.sub(r'\s+', ' ', buf).strip()


def notisbn(isbnlike, level='strict'):
    """Check with the goal to invalidate isbn-like.

    level:
    'strict' when certain they are not ISBNs (default)
    'loose'  only filters obvious NO ISBNs

    """
    if level not in ('strict', 'loose'):  # pragma: no cover
        LOGGER.error('level as no option %s', level)
        return None
    isbnlike = canonical(isbnlike)
    if len(isbnlike) not in (10, 13):
        return True
    if level != 'strict':
        return False
    if len(isbnlike) == 10:
        return not is_isbn10(isbnlike)
    return not is_isbn13(isbnlike)


def get_isbnlike(text, level='normal'):
    """Extract all substrings that seem like ISBNs.

    level:
    strict almost as certain they are ISBNs
    normal (default)
    loose  catch many as possible

    """
    if level == 'normal':  # pragma: no cover
        isbnlike = RE_NORMAL
    elif level == 'strict':
        isbnlike = RE_STRICT
    elif level == 'loose':
        isbnlike = RE_LOOSE
    else:
        LOGGER.error('level as no option %s', level)
        return []
    return isbnlike.findall(text)


def get_canonical_isbn(isbnlike, output='bouth'):
    """Extract ISBNs and transform them to the canonical form.

    output:
    isbn10
    isbn13
    bouth  (default)

    """
    if output not in ('bouth', 'isbn10', 'isbn13'):  # pragma: no cover
        LOGGER.error('output as no option %s', output)
        return ''

    regex = RE_NORMAL

    match = regex.search(isbnlike)
    if match:
        # Get only canonical characters
        cisbn = canonical(match.group())
        if not cisbn:
            return ''
        # Split into a list
        chars = list(cisbn)
        # Remove the last digit from `chars` and assign it to `last`
        last = chars.pop()
        buf = ''.join(chars)

        if len(chars) == 9:
            # Compute the ISBN-10 checksum digit
            check = check_digit10(buf)
        else:
            # Compute the ISBN-13 checksum digit
            check = check_digit13(buf)

        # If checksum OK return a `canonical` ISBN
        if str(check) == last:
            if output == 'bouth':
                return cisbn
            if output == 'isbn10':
                return cisbn if len(cisbn) == 10 else to_isbn10(cisbn)
            return to_isbn13(cisbn) if len(cisbn) == 10 else cisbn
    return ''


def ean13(isbnlike):
    """Transform an `isbnlike` string in an EAN number (canonical ISBN-13)."""
    ib = canonical(isbnlike)
    if len(ib) == 13:
        return ib if is_isbn13(ib) else ''
    if len(ib) == 10:
        return to_isbn13(ib) if is_isbn10(ib) else ''
    return ''


# Alias
EAN13 = ean13
GTIN13 = ean13
